src/dbnode/encoding/proto/iterator.go

```diff
@@ -0,0 +1,878 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/m3db/m3x/checked"
+
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3/src/dbnode/ts"
+	xtime "github.com/m3db/m3x/time"
+)
+
+const (
+	// Maximum capacity of a checked.Bytes that will be retained between resets.
+	maxCapacityUnmarshalBufferRetain = 1024
+)
+
+var (
+	itErrPrefix                 = "proto iterator:"
+	errIteratorReaderIsRequired = fmt.Errorf("%s reader is required", itErrPrefix)
+	errIteratorSchemaIsRequired = fmt.Errorf("%s schema is required", itErrPrefix)
+)
+
+type iterator struct {
+	opts                   encoding.Options
+	err                    error
+	schema                 *desc.MessageDescriptor
+	stream                 encoding.IStream
+	consumedFirstMessage   bool
+	lastIterated           *dynamic.Message
+	lastIteratedAnnotation []byte
```

Comment:
Why name this Annotation, isn't it just the proto bytes now and we're just using annotation due to interface refactoring issues?

Reply:
sure I can rename this